src/disc_bsd.c

```diff
@@ -0,0 +1,226 @@
+			    unsigned int features) {
+	mb_disc_toc toc;
+	char device_name[MAX_DEV_LEN] = "";
+	int device_number;
+	int fd;
+
+	device_number = (int) strtol(device, NULL, 10);
+
+	if (device_number > 0) {
+		if(!get_device(device_number, device_name, MAX_DEV_LEN)) {
+			snprintf(disc->error_msg, MB_ERROR_MSG_LENGTH,
+				 "cannot find cd device with the number '%d'",
+				 device_number);
+			return 0; /* error */
+		}
+		device = device_name;
+	}
+
+	fd = mb_disc_unix_open(disc, device);
+	if (fd < 0)
+		return 0;
+
+	{
+		struct ioc_toc_header th;
+
+		if (ioctl(fd, CDIOREADTOCHEADER, &th) < 0) {
+			snprintf(disc->error_msg, MB_ERROR_MSG_LENGTH,
+				 "cannot read table of contents");
+			close(fd);
+			return 0; /* error */
+		}
+
+		toc.first_track_num = th.starting_track;
+		toc.last_track_num  = th.ending_track;
+
+		if (toc.last_track_num == 0) {
+			snprintf(disc->error_msg, MB_ERROR_MSG_LENGTH,
+				 "this disc has no tracks");
+			close(fd);
+			return 0;
+		}
+	}
+	{
+		struct cd_toc_entry te[100];
+		struct ioc_read_toc_entry rte;
+		int i;
+
+		memset(&te,  0, sizeof  te);
+		memset(&rte, 0, sizeof rte);
+		rte.address_format = CD_LBA_FORMAT;
+		rte.data           = &te[0];
+		rte.data_len       = sizeof te;
+		rte.starting_track = toc.first_track_num;
+
+		if (ioctl(fd, CDIOREADTOCENTRYS, &rte) < 0) {
+			snprintf(disc->error_msg, MB_ERROR_MSG_LENGTH,
+				 "cannot read table of contents");
+			close(fd);
+			return 0; /* error */
+		}
+
+		for (i = toc.first_track_num; i <= toc.last_track_num; ++i) {
+			assert (te[i - toc.first_track_num].track == i);
+#if defined(__FreeBSD__) /* LBA address is in network byte order */
+			toc.tracks[i].address = ntohl(te[i - toc.first_track_num].addr.lba);
+#else
+			toc.tracks[i].address = te[i - toc.first_track_num].addr.lba;
+#endif
+			toc.tracks[i].control = te[i - toc.first_track_num].control;
+		}
+		/* leadout - track number 170 (0xAA) */
+		assert (te[i - toc.first_track_num].track == 0xAA);
+#if defined(__FreeBSD__) /* LBA address is in network byte order */
+		toc.tracks[0].address = ntohl(te[i - toc.first_track_num].addr.lba);
+#else
+		toc.tracks[0].address = te[i - toc.first_track_num].addr.lba;
+#endif
+		toc.tracks[0].control = te[i - toc.first_track_num].control;
+	}
+
+	if ( !mb_disc_load_toc(disc, &toc) )
+		return 0;
+
+	/* Read in the media catalog number */
+	if ((features & DISCID_FEATURE_MCN) != 0
+		&& mb_disc_has_feature_unportable(DISCID_FEATURE_MCN)) {
+		mb_disc_unix_read_mcn(fd, disc);
+	}
+
+	/* Read the ISRC for the track */
+	if ((features & DISCID_FEATURE_ISRC) != 0
+		&& mb_disc_has_feature_unportable(DISCID_FEATURE_ISRC)) {
+		int i;
+		for (i = disc->first_track_num; i <= disc->last_track_num; ++i) {
+			mb_disc_unix_read_isrc(fd, disc, i);
+		}
+	}
+
+	close(fd);
+
+	return 1;
+}
+
+char *mb_disc_get_default_device_unportable(void) {
+	static char result[MAX_DEV_LEN + 1];
+	return get_device(1, result, sizeof result) ? result : NULL;
```

Comment:
I would prefer defining MB_DEFAULT_DEVICE as /dev/cd0 or /dev/rcd0c depending on the os, like it was before ;-)
and then returning MB_DEFAULT_DEVICE in case get_device "fails".